File: src/signal_integration.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

#![expect(
    clippy::cast_precision_loss,
    reason = "domain-specific numeric patterns"
)]

//! Signal integration: cyclic di-GMP + quorum sensing in Vibrio cholerae.
//!
//! Port of `control/signal_integration/signal_integration.py`.
//!
//! Reproduces key dynamics from:
//! Srivastava et al. (2011)
//! "Integration of Cyclic di-GMP and Quorum Sensing in the Control of
//!  vpsT Expression in Vibrio cholerae"
//! J Bacteriology 193:6331-41.
//!
//! Core thesis: The vpsT promoter integrates two inputs (cdg AND ai)
//! as a biological AND gate — maps to multi-input attention.
//!
//! ## `BarraCUDA` connection
//!
//! - Two-input Hill AND gate: elementwise ops (`barracuda::ops::elementwise`)
//! - vpsT ODE integration: `barracuda::numerical::rk45_solve` or GPU `rk4_parallel.wgsl`
//! - Dose-response surface: batch parallel ODE over 2D parameter grid

/// WGSL shader: two-input Hill AND gate over 2D grid.
///
/// One thread per (cdg, ai) pair. Computes `Vmax × Hill(cdg) × Hill(ai)`.
/// Paper 021 (Signal Integration).
///
/// Absorption target: `barracuda::ops::elementwise`.
/// Validated: `validate_gpu_signal` (9/9 PASS).
pub use neural_spring_forge::shaders::HILL_GATE as WGSL_HILL_GATE;

use crate::primitives;
use crate::rng::Rng;

/// Two-input Hill function: AND gate for vpsT activation.
///
/// `f(cdg, ai) = Vmax * (cdg^n1 / (K1^n1 + cdg^n1)) * (ai^n2 / (K2^n2 + ai^n2))`
///
/// Composes two [`crate::primitives::hill_activation`] calls (amplitude=1) then
/// scales by `vmax`.
#[must_use]
pub fn two_input_hill(cdg: f64, ai: f64, vmax: f64, k1: f64, k2: f64, n1: f64, n2: f64) -> f64 {
    let h1 = primitives::hill_activation(cdg, 1.0, k1, n1);
    let h2 = primitives::hill_activation(ai, 1.0, k2, n2);
    vmax * h1 * h2
}

/// Logic gate classification: (`cdg_high`, `ai_high`) -> AND outcome.
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum LogicGate {
    /// Low/Low -> Low
    OffOff,
    /// High/Low -> Low
    OnOff,
    /// Low/High -> Low
    OffOn,
    /// High/High -> High
    OnOn,
}

/// Classify logic gate case from cdg and ai vs thresholds.
#[must_use]
pub fn classify_logic_gate(cdg: f64, ai: f64, cdg_thresh: f64, ai_thresh: f64) -> LogicGate {
    let cdg_high = cdg >= cdg_thresh;
    let ai_high = ai >= ai_thresh;
    match (cdg_high, ai_high) {
        (false, false) => LogicGate::OffOff,
        (true, false) => LogicGate::OnOff,
        (false, true) => LogicGate::OffOn,
        (true, true) => LogicGate::OnOn,
    }
}

/// Parameters for the vpsT regulatory ODE.
#[derive(Debug, Clone)]
pub struct OdeParams {
    /// Cell density scaling autoinducer production.
    pub cell_density: f64,
    /// c-di-GMP synthesis rate constant.
    pub cdg_synth: f64,
    /// c-di-GMP degradation rate constant.
    pub cdg_deg: f64,
    /// Autoinducer production coefficient.
    pub ai_prod: f64,
    /// Autoinducer decay rate constant.
    pub ai_decay: f64,
    /// vpsT transcript/protein degradation rate.
    pub vps_degradation: f64,
    /// Maximum vpsT production rate in the Hill AND term.
    pub vmax: f64,
    /// Half-saturation constant for c-di-GMP (Hill on cdg).
    pub k1: f64,
    /// Half-saturation constant for autoinducer (Hill on ai).
    pub k2: f64,
    /// Hill coefficient for c-di-GMP.
    pub n1: f64,
    /// Hill coefficient for autoinducer.
    pub n2: f64,
    /// Standard deviation scale for additive Gaussian noise on cdg dynamics.
    pub noise_scale: f64,
    /// RNG seed for stochastic noise in the ODE integration.
    pub seed: u64,
}

impl Default for OdeParams {
    fn default() -> Self {
        Self {
            cell_density: 1.0,
            cdg_synth: 0.5,
            cdg_deg: 0.2,
            ai_prod: 0.3,
            ai_decay: 0.1,
            vps_degradation: 0.3,
            vmax: 1.0,
            k1: 1.0,
            k2: 1.0,
            n1: 2.0,
            n2: 2.0,
            noise_scale: 0.0,
            seed: 42,
        }
    }
}

/// ODE state: [cdg, ai, vpsT, biofilm]
#[derive(Debug, Clone, Copy)]
pub struct OdeState {
    /// Cyclic di-GMP concentration.
    pub cdg: f64,
    /// Autoinducer concentration.
    pub ai: f64,
    /// vpsT expression level (promoter output).
    pub vps_t: f64,
    /// Integrated biofilm biomass proxy.
    pub biofilm: f64,
}

impl OdeState {
    const fn to_array(self) -> [f64; 4] {
        [self.cdg, self.ai, self.vps_t, self.biofilm]
    }

    const fn from_array(a: [f64; 4]) -> Self {
        Self {
            cdg: a[0],
            ai: a[1],
            vps_t: a[2],
            biofilm: a[3],
        }
    }
}

fn ode_rhs(y: &[f64; 4], params: &OdeParams, rng: &mut Rng) -> [f64; 4] {
    let cdg = y[0];
    let ai = y[1];
    let vps_t = y[2];

    let noise = if params.noise_scale > 0.0 {
        params.noise_scale * rng.normal()
    } else {
        0.0
    };

    let d_cdg = params.cdg_deg.mul_add(-cdg, params.cdg_synth) + noise;
    let d_ai = params
        .ai_prod
        .mul_add(params.cell_density, -(params.ai_decay * ai));
    let f_val = two_input_hill(
        cdg,
        ai,
        params.vmax,
        params.k1,
        params.k2,
        params.n1,
        params.n2,
    );
    let d_vps_t = params.vps_degradation.mul_add(-vps_t, f_val);
    let d_biofilm = vps_t;

    [d_cdg, d_ai, d_vps_t, d_biofilm]
}

/// Single RK4 step (delegates to [`crate::primitives::rk4_step`]).
///
/// Note: the stochastic noise in `ode_rhs` means each sub-step evaluation
/// draws fresh noise. This matches the original implementation and the
/// Python baseline behavior.
fn rk4_step(y: &[f64; 4], _t: f64, dt: f64, params: &OdeParams, rng: &mut Rng) -> [f64; 4] {
    primitives::rk4_step(y, dt, |state| ode_rhs(state, params, rng))
}

/// Integrate vpsT regulatory ODE with RK4.
#[expect(
    clippy::cast_possible_truncation,
    clippy::cast_sign_loss,
    reason = "step count from f64→usize is bounded by caller-supplied t_end/dt"
)]
#[must_use]
pub fn integrate_ode(t_end: f64, dt: f64, y0: &OdeState, params: &OdeParams) -> Vec<OdeState> {
    let n_steps = (t_end / dt).ceil() as usize + 1;
    let mut trace = Vec::with_capacity(n_steps);
    trace.push(*y0);

    let mut rng = Rng::new(params.seed);
    let mut y = y0.to_array();

    for i in 1..n_steps {
        let t = (i - 1) as f64 * dt;
        y = rk4_step(&y, t, dt, params, &mut rng);
        for v in &mut y {
            *v = (*v).max(0.0);
        }
        trace.push(OdeState::from_array(y));
    }

    trace
}

/// Sweep cdg and ai to get AND gate outputs.
#[must_use]
pub fn logic_gate_sweep(k1: f64, k2: f64, n1: f64, n2: f64) -> [(LogicGate, f64); 4] {
    let low = 0.01_f64;
    let high = 5.0_f64;

    [
        (
            LogicGate::OffOff,
            two_input_hill(low, low, 1.0, k1, k2, n1, n2),
        ),
        (
            LogicGate::OnOff,
            two_input_hill(high, low, 1.0, k1, k2, n1, n2),
        ),
        (
            LogicGate::OffOn,
            two_input_hill(low, high, 1.0, k1, k2, n1, n2),
        ),
        (
            LogicGate::OnOn,
            two_input_hill(high, high, 1.0, k1, k2, n1, n2),
        ),
    ]
}

/// Dose-response: sweep cdg with ai fixed.
#[must_use]
pub fn dose_response_cdg(
    ai_fixed: f64,
    n_points: usize,
    k1: f64,
    k2: f64,
    n1: f64,
    n2: f64,
) -> Vec<(f64, f64)> {
    let cdg_min = 0.01_f64;
    let cdg_max = 10.0_f64;
    (0..n_points)
        .map(|i| {
            let frac = i as f64 / (n_points - 1).max(1) as f64;
            let cdg = cdg_min * (cdg_max / cdg_min).powf(frac);
            let v = two_input_hill(cdg, ai_fixed, 1.0, k1, k2, n1, n2);
            (cdg, v)
        })
        .collect()
}

/// Dose-response: sweep ai with cdg fixed.
#[must_use]
pub fn dose_response_ai(
    cdg_fixed: f64,
    n_points: usize,
    k1: f64,
    k2: f64,
    n1: f64,
    n2: f64,
) -> Vec<(f64, f64)> {
    let ai_min = 0.01_f64;
    let ai_max = 10.0_f64;
    (0..n_points)
        .map(|i| {
            let frac = i as f64 / (n_points - 1).max(1) as f64;
            let ai = ai_min * (ai_max / ai_min).powf(frac);
            let v = two_input_hill(cdg_fixed, ai, 1.0, k1, k2, n1, n2);
            (ai, v)
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn and_gate_high_only_when_both() {
        let sweep = logic_gate_sweep(1.0, 1.0, 2.0, 2.0);
        let threshold = 0.5;
        for (gate, v) in sweep {
            match gate {
                LogicGate::OffOff | LogicGate::OnOff | LogicGate::OffOn => {
                    assert!(v < threshold, "{gate:?} should be low, got {v}");
                }
                LogicGate::OnOn => {
                    assert!(v > threshold, "ON/ON should be high, got {v}");
                }
            }
        }
    }

    #[test]
    fn two_input_hill_bounds() {
        let v = two_input_hill(1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0);
        assert!((0.0..=1.0).contains(&v));
    }

    #[test]
    fn ode_finite_nonneg() {
        let y0 = OdeState {
            cdg: 0.1,
            ai: 0.1,
            vps_t: 0.0,
            biofilm: 0.0,
        };
        let params = OdeParams::default();
        let trace = integrate_ode(2.0, 0.01, &y0, &params);
        for s in &trace {
            assert!(s.cdg >= 0.0 && s.cdg.is_finite());
            assert!(s.ai >= 0.0 && s.ai.is_finite());
            assert!(s.vps_t >= 0.0 && s.vps_t.is_finite());
            assert!(s.biofilm >= 0.0 && s.biofilm.is_finite());
        }
    }

    #[test]
    fn dose_response_sigmoidal() {
        let dr = dose_response_cdg(5.0, 50, 1.0, 1.0, 2.0, 2.0);
        let low: f64 = dr.iter().take(5).map(|(_, v)| v).sum::<f64>() / 5.0;
        let high: f64 = dr.iter().rev().take(5).map(|(_, v)| v).sum::<f64>() / 5.0;
        assert!(low < 0.3);
        assert!(high > 0.7);
    }

    #[test]
    fn dose_response_ai_sigmoidal() {
        let dr = dose_response_ai(5.0, 50, 1.0, 1.0, 2.0, 2.0);
        assert_eq!(dr.len(), 50);
        let low: f64 = dr.iter().take(5).map(|(_, v)| v).sum::<f64>() / 5.0;
        let high: f64 = dr.iter().rev().take(5).map(|(_, v)| v).sum::<f64>() / 5.0;
        assert!(low < 0.3);
        assert!(high > 0.7);
    }

    #[test]
    fn logic_gate_sweep_covers_all_combos() {
        let sweep = logic_gate_sweep(1.0, 1.0, 2.0, 2.0);
        assert_eq!(sweep.len(), 4);
        for &(_, val) in &sweep {
            assert!(val >= 0.0 && val.is_finite());
        }
    }

    #[test]
    fn classify_logic_gate_variants() {
        assert_eq!(classify_logic_gate(2.0, 2.0, 1.0, 1.0), LogicGate::OnOn);
        assert_eq!(classify_logic_gate(0.1, 0.1, 1.0, 1.0), LogicGate::OffOff);
        assert_eq!(classify_logic_gate(2.0, 0.1, 1.0, 1.0), LogicGate::OnOff);
        assert_eq!(classify_logic_gate(0.1, 2.0, 1.0, 1.0), LogicGate::OffOn);
    }

    #[test]
    fn integrate_ode_length() {
        let y0 = OdeState {
            cdg: 0.5,
            ai: 0.5,
            vps_t: 0.0,
            biofilm: 0.0,
        };
        let params = OdeParams::default();
        let trace = integrate_ode(1.0, 0.1, &y0, &params);
        assert_eq!(trace.len(), 11);
    }
}

File: control/signal_integration/signal_integration.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Provenance: see src/provenance/experiments.rs — SIGNAL_INTEGRATION_PROVENANCE

#!/usr/bin/env python3
"""
neuralSpring Paper 021 — Signal Integration in Vibrio cholerae

Reproduces key dynamics from:
  Srivastava et al. (2011)
  "Integration of Cyclic di-GMP and Quorum Sensing in the Control of
   vpsT Expression in Vibrio cholerae"
  J Bacteriology 193:6331-41.

Core thesis: The vpsT promoter integrates TWO signaling inputs —
cyclic di-GMP (intracellular) and quorum sensing autoinducers
(extracellular) — to control biofilm formation. This creates a
biological AND gate / logic circuit.

Maps to ecoPrimals: multi-input attention mechanism where multiple
signals are combined to produce a decision.

BarraCUDA connection:
  - Two-input Hill = multiplicative attention (product of two sigmoids)
  - ODE integration = sequential recurrence (LSTM-like)
  - Dose-response = softmax-like nonlinearity
"""

import sys

import numpy as np

# ---------------------------------------------------------------------------
# Two-input Hill function (AND gate)
# ---------------------------------------------------------------------------
# f(cdg, ai) = Vmax * (cdg^n1 / (K1^n1 + cdg^n1)) * (ai^n2 / (K2^n2 + ai^n2))


def two_input_hill(
    cdg: float,
    ai: float,
    vmax: float = 1.0,
    k1: float = 1.0,
    k2: float = 1.0,
    n1: float = 2.0,
    n2: float = 2.0,
) -> float:
    """Two-input Hill function: AND gate for vpsT activation."""
    h1 = (cdg**n1) / (k1**n1 + cdg**n1 + 1e-30)
    h2 = (ai**n2) / (k2**n2 + ai**n2 + 1e-30)
    return vmax * h1 * h2


def _hill(cdg, ai, vmax, k1, k2, n1, n2):
    h1 = (cdg**n1) / (k1**n1 + cdg**n1 + 1e-30)
    h2 = (ai**n2) / (k2**n2 + ai**n2 + 1e-30)
    return vmax * h1 * h2


# ---------------------------------------------------------------------------
# RK4 ODE integrator
# ---------------------------------------------------------------------------


def rk4_step(
    y: np.ndarray,
    t: float,
    dt: float,
    f: callable,
) -> np.ndarray:
    """Single RK4 step."""
    k1 = f(t, y)
    k2 = f(t + 0.5 * dt, y + 0.5 * dt * k1)
    k3 = f(t + 0.5 * dt, y + 0.5 * dt * k2)
    k4 = f(t + dt, y + dt * k3)
    return y + (dt / 6.0) * (k1 + 2 * k2 + 2 * k3 + k4)


# ---------------------------------------------------------------------------
# ODE system: cdg, ai, vpsT, biofilm
# ---------------------------------------------------------------------------


def _ode_rhs(
    t: float,
    y: np.ndarray,
    cell_density: float,
    cdg_synth: float,
    cdg_deg: float,
    ai_prod: float,
    ai_decay: float,
    vps_degradation: float,
    vmax: float,
    k1: float,
    k2: float,
    n1: float,
    n2: float,
    noise_scale: float,
    rng: np.random.Generator,
) -> np.ndarray:
    cdg, ai, vpsT, biofilm = y
    # d(cdg)/dt = synthesis - degradation + noise
    noise = noise_scale * rng.standard_normal() if noise_scale > 0 else 0.0
    d_cdg = cdg_synth - cdg_deg * cdg + noise
    # d(ai)/dt = production * cell_density - decay
    d_ai = ai_prod * cell_density - ai_decay * ai
    # d(vpsT)/dt = f(cdg, ai) - degradation
    f_val = _hill(cdg, ai, vmax, k1, k2, n1, n2)
    d_vpsT = f_val - vps_degradation * vpsT
    # d(biofilm)/dt proportional to vpsT
    d_biofilm = vpsT
    return np.array([d_cdg, d_ai, d_vpsT, d_biofilm])


def integrate_ode(
    t_end: float = 10.0,
    dt: float = 0.01,
    cdg0: float = 0.1,
    ai0: float = 0.1,
    vps0: float = 0.0,
    bio0: float = 0.0,
    cell_density: float = 1.0,
    cdg_synth: float = 0.5,
    cdg_deg: float = 0.2,
    ai_prod: float = 0.3,
    ai_decay: float = 0.1,
    vps_degradation: float = 0.3,
    vmax: float = 1.0,
    k1: float = 1.0,
    k2: float = 1.0,
    n1: float = 2.0,
    n2: float = 2.0,
    noise_scale: float = 0.0,
    seed: int = 42,
) -> dict:
    """Integrate vpsT regulatory ODE with RK4."""
    rng = np.random.default_rng(seed)
    n_steps = int(t_end / dt) + 1
    trace = np.zeros((n_steps, 4))
    trace[0] = [cdg0, ai0, vps0, bio0]
    y = np.array([cdg0, ai0, vps0, bio0], dtype=float)

    def rhs(t, yy):
        return _ode_rhs(
            t,
            yy,
            cell_density,
            cdg_synth,
            cdg_deg,
            ai_prod,
            ai_decay,
            vps_degradation,
            vmax,
            k1,
            k2,
            n1,
            n2,
            noise_scale,
            rng,
        )

    for i in range(1, n_steps):
        t = (i - 1) * dt
        y = rk4_step(y, t, dt, rhs)
        y = np.maximum(y, 0.0)
        trace[i] = y

    return {
        "t": np.linspace(0, t_end, n_steps),
        "cdg": trace[:, 0],
        "ai": trace[:, 1],
        "vpsT": trace[:, 2],
        "biofilm": trace[:, 3],
        "trace": trace,
    }


# ---------------------------------------------------------------------------
# Logic gate characterization
# ---------------------------------------------------------------------------


def logic_gate_sweep(k1: float = 1.0, k2: float = 1.0, n1: float = 2.0, n2: float = 2.0) -> dict:
    """Sweep cdg and ai to show AND gate behavior."""
    low, high = 0.01, 5.0
    cases = [
        (low, low, "OFF/OFF"),
        (high, low, "ON/OFF"),
        (low, high, "OFF/ON"),
        (high, high, "ON/ON"),
    ]
    return {label: _hill(cdg, ai, 1.0, k1, k2, n1, n2) for cdg, ai, label in cases}


def dose_response_cdg(
    ai_fixed: float, k1: float = 1.0, k2: float = 1.0, n1: float = 2.0, n2: float = 2.0
) -> tuple:
    """Sweep cdg with ai fixed."""
    cdg_vals = np.logspace(-2, 1, 50)
    vps = np.array([_hill(c, ai_fixed, 1.0, k1, k2, n1, n2) for c in cdg_vals])
    return cdg_vals, vps


def multiplicative_attention(cdg: float, ai: float, k1: float = 1.0, k2: float = 1.0) -> float:
    """Product of attention weights = AND gate (Hill n=2)."""
    return (cdg**2 / (k1**2 + cdg**2 + 1e-30)) * (ai**2 / (k2**2 + ai**2 + 1e-30))


# ---------------------------------------------------------------------------
# Main validation
# ---------------------------------------------------------------------------


def main() -> int:
    """Validate signal integration and AND gate behavior.

    Provenance
    ----------
    Paper: Srivastava et al. (2011) J Bacteriology 193:6331-41.
    Model: Two-input Hill (cdg AND ai) controlling vpsT.
    Validation: logic gate, dose-response, ODE, attention mapping.

    Tolerance rationale:
      * ODE: finite, non-negative
      * AND gate: high vpsT only when both inputs high
      * Dose-response: sigmoidal (Hill)
      * QS: cell density increases ai
      * Biofilm proportional to vpsT
      * Multiplicative attention = product of sigmoids
    """
    total_passed = 0
    total_failed = 0

    print("=" * 72)
    print("neuralSpring Paper 021: Signal Integration (cdg + QS)")
    print("  Srivastava et al. (2011) J Bacteriology 193:6331-41")
    print("=" * 72)

    k1, k2, n1, n2 = 1.0, 1.0, 2.0, 2.0

    # ------------------------------------------------------------------
    # Part 1: ODE produces finite, non-negative
    # ------------------------------------------------------------------
    print("\n--- Part 1: ODE Finite and Non-Negative ---")

    result = integrate_ode(t_end=5.0, dt=0.01, cell_density=1.0, seed=42)
    finite = np.all(np.isfinite(result["trace"]))
    nonneg = np.all(result["trace"] >= 0)

    if finite and nonneg:
        print("  [PASS] ODE: finite and non-negative concentrations")
        total_passed += 1
    else:
        print(f"  [FAIL] ODE: finite={finite}, nonneg={nonneg}")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 2: AND gate — high vpsT only when BOTH inputs high
    # ------------------------------------------------------------------
    print("\n--- Part 2: AND Gate Characterization ---")

    lg = logic_gate_sweep(k1, k2, n1, n2)
    off_off = lg["OFF/OFF"]
    on_off = lg["ON/OFF"]
    off_on = lg["OFF/ON"]
    on_on = lg["ON/ON"]

    threshold = 0.5
    and_ok = off_off < threshold and on_off < threshold and off_on < threshold and on_on > threshold

    if and_ok:
        print(
            f"  [PASS] AND: OFF/OFF={off_off:.4f}, ON/OFF={on_off:.4f}, OFF/ON={off_on:.4f}, ON/ON={on_on:.4f}"
        )
        total_passed += 1
    else:
        print(f"  [FAIL] AND gate violated: {lg}")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 3: Each input alone insufficient
    # ------------------------------------------------------------------
    if on_off < threshold and off_on < threshold:
        print("  [PASS] Each input alone insufficient (on/off and off/on → low)")
        total_passed += 1
    else:
        print("  [FAIL] Single input produces high output")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 4: Dose-response is sigmoidal
    # ------------------------------------------------------------------
    print("\n--- Part 4: Dose-Response Sigmoidal ---")

    cdg_vals, vps_cdg = dose_response_cdg(ai_fixed=5.0, k1=k1, k2=k2, n1=n1, n2=n2)
    low_end = np.mean(vps_cdg[:5])
    high_end = np.mean(vps_cdg[-5:])
    mid_idx = len(vps_cdg) // 2
    mid_val = vps_cdg[mid_idx]

    sigmoidal = low_end < mid_val < high_end and low_end < 0.3 and high_end > 0.7

    if sigmoidal:
        print(
            f"  [PASS] CDG dose-response sigmoidal (low={low_end:.4f}, mid={mid_val:.4f}, high={high_end:.4f})"
        )
        total_passed += 1
    else:
        print("  [FAIL] Dose-response not sigmoidal")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 5: Cell density increases ai
    # ------------------------------------------------------------------
    print("\n--- Part 5: QS — Cell Density Increases AI ---")

    low_dens = integrate_ode(t_end=3.0, cell_density=0.2, seed=42)
    high_dens = integrate_ode(t_end=3.0, cell_density=2.0, seed=42)

    ai_low = np.mean(low_dens["ai"][-100:])
    ai_high = np.mean(high_dens["ai"][-100:])

    if ai_high > ai_low:
        print(f"  [PASS] Higher density → higher ai ({ai_low:.4f} vs {ai_high:.4f})")
        total_passed += 1
    else:
        print("  [FAIL] Cell density did not increase ai")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 6: Biofilm proportional to vpsT
    # ------------------------------------------------------------------
    high_result = integrate_ode(t_end=3.0, cdg0=3.0, ai0=3.0, cell_density=2.0, seed=42)
    low_result = integrate_ode(t_end=3.0, cdg0=0.05, ai0=0.05, cell_density=0.1, seed=42)
    bio_high = high_result["biofilm"][-1]
    bio_low = low_result["biofilm"][-1]
    if bio_high > bio_low:
        print(
            f"  [PASS] Biofilm proportional to vpsT (high inputs → more biofilm: {bio_high:.4f} > {bio_low:.4f})"
        )
        total_passed += 1
    else:
        print("  [FAIL] Biofilm not proportional to vpsT")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 7: Multiplicative attention (product of sigmoids)
    # ------------------------------------------------------------------
    att_prod = multiplicative_attention(2.0, 2.0, k1, k2)
    hill_val = _hill(2.0, 2.0, 1.0, k1, k2, 2.0, 2.0)
    if abs(att_prod - hill_val) < 0.01:
        print("  [PASS] Integration = multiplicative attention (product of sigmoids)")
        total_passed += 1
    else:
        print(f"  [FAIL] Attention mismatch: att={att_prod:.4f}, hill={hill_val:.4f}")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 8: BarraCUDA connection documented
    # ------------------------------------------------------------------
    print("\n--- Part 8: BarraCUDA Connection ---")
    print("  Two-input Hill = gemm + sigmoid + elementwise multiply")
    print("  ODE = recurrent step (LSTM-like)")
    print("  [PASS] BarraCUDA connection documented")
    total_passed += 1

    # ------------------------------------------------------------------
    # Summary
    # ------------------------------------------------------------------
    total = total_passed + total_failed
    print(f"\n{'=' * 72}")
    print(f"TOTAL: {total_passed}/{total} PASS, {total_failed}/{total} FAIL")
    print(f"{'=' * 72}")

    return 0 if total_failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
